core/src/main/java/org/apache/iceberg/BaseIncrementalAppendScan.java

```diff
@@ -0,0 +1,179 @@
+    return newRefinedScan(tableOps(), table(), schema(), context().fromSnapshotIdInclusive(fromSnapshotId));
+  }
+
+  @Override
+  public IncrementalAppendScan fromSnapshotExclusive(long fromSnapshotId) {
+    // for exclusive behavior, table().snapshot(fromSnapshotId) check can't be applied.
+    // as fromSnapshotId could be matched to a parent snapshot that is already expired
+    return newRefinedScan(tableOps(), table(), schema(), context().fromSnapshotId(fromSnapshotId));
+  }
+
+  @Override
+  public IncrementalAppendScan toSnapshot(long toSnapshotId) {
+    Preconditions.checkArgument(table().snapshot(toSnapshotId) != null,
+        "Cannot find end snapshot: %d", toSnapshotId);
+    return newRefinedScan(tableOps(), table(), schema(),  context().toSnapshotId(toSnapshotId));
+  }
+
+  @Override
+  public CloseableIterable<FileScanTask> planFiles() {
+    long toSnapshotIdInclusive;
+    if (context().toSnapshotId() != null) {
+      toSnapshotIdInclusive = context().toSnapshotId();
+    } else if (table().currentSnapshot() != null) {
+      toSnapshotIdInclusive = table().currentSnapshot().snapshotId();
+    } else if (context().fromSnapshotId() != null) {
+      throw new IllegalArgumentException(
+          "Invalid config: end snapshot is not set, start snapshot is set, and table has no current snapshot");
+    } else {
+      // It is an empty table (no current snapshot). Both from and to snapshots aren't set either.
+      // In this case, listener notification is also skipped
+      return CloseableIterable.empty();
+    }
+
+    // fromSnapshotIdExclusive can be null. appendsBetween handles null fromSnapshotIdExclusive properly
+    // by finding the oldest ancestor of end snapshot.
+    Long fromSnapshotIdExclusive = context().fromSnapshotId();
+    if (context().fromSnapshotId() != null) {
+      if (context().fromSnapshotInclusive()) {
+        // validate the fromSnapshotId is an ancestor of toSnapshotId
+        Preconditions.checkArgument(
+            SnapshotUtil.isAncestorOf(table(), toSnapshotIdInclusive, context().fromSnapshotId()),
+            "Starting snapshot (inclusive) %s is not an ancestor of end snapshot %s",
+            context().fromSnapshotId(), toSnapshotIdInclusive);
+        // for inclusive behavior fromSnapshotIdExclusive is set to the parent snapshot id, which can be null.
+        fromSnapshotIdExclusive = table().snapshot(context().fromSnapshotId()).parentId();
+      } else {
+        // validate the parent snapshot id an ancestor of toSnapshotId
+        Preconditions.checkArgument(
+            SnapshotUtil.isParentAncestorOf(table(), toSnapshotIdInclusive, context().fromSnapshotId()),
+            "Starting snapshot (exclusive) %s is not a parent ancestor of end snapshot %s",
+            context().fromSnapshotId(), toSnapshotIdInclusive);
+      }
+    }
+
+    if (fromSnapshotIdExclusive != null) {
+      Listeners.notifyAll(new IncrementalScanEvent(table().name(), fromSnapshotIdExclusive,
```

Comment:
nit: The purpose of the true/false flags is not clear without looking into the code. Would it make sense to offer static methods like `fromSnapshotInclusive` and `fromSnapshotExclusive` in `IncrementalScanEvent`?

Reply:
@aokolnychyi agree with your comment that boolean flag is not clear without looking into the code. Just static methods like `fromSnapshotInclusive` and `fromSnapshotExclusive` is also a little odd to me. Do you think we should provide a builder class instead? Then in the builder class, we can provide methods like `fromSnapshotInclusive` and `fromSnapshotExclusive` methods.

Reply:
I mentioned separate methods as we usually prefer a named method to give some meaning to a boolean flag but I agree there are quite some args in this case so such it is questionable. Builder sounds good but this point is minor. Up to you. I think we can leave as-is and don't worry about it.

Reply:
I like the builder slightly better. but as you said, this is minor. We can leave as-is for now.